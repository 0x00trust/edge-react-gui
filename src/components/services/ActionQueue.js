// @flow

import { type EdgeAccount } from 'edge-core-js'

import { makeActionQueueStore } from '../../controllers/action-queue/ActionQueueStore'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useDispatch, useSelector } from '../../types/reactRedux'

export const ActionQueue = () => {
  const dispatch = useDispatch()
  const account: EdgeAccount = useSelector(state => state.core.account)

  //
  // Initialization
  //

  useAsyncEffect(async () => {
    if (account?.dataStore != null) {
      const store = makeActionQueueStore(account)
      const queue = await store.getActionQueueMap()
      dispatch({
        type: 'ACTION_QUEUE/LOAD_QUEUE',
        data: queue
      })
    }
  }, [account, dispatch])

  return null
}
