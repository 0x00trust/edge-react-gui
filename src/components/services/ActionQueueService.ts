import { EdgeAccount } from 'edge-core-js'
import * as React from 'react'

import { makeActionQueueStore } from '../../controllers/action-queue/ActionQueueStore'
import { mockActionProgram } from '../../controllers/action-queue/mock'
import { updateActionProgramState } from '../../controllers/action-queue/redux/actions'
import { executeActionProgram } from '../../controllers/action-queue/runtime'
import { ActionQueueMap, ExecutionContext } from '../../controllers/action-queue/types'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useMemo } from '../../types/reactHooks'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { makePeriodicTask } from '../../util/PeriodicTask'

const EXECUTION_INTERVAL = 1000

export const ActionQueueService = () => {
  const dispatch = useDispatch()
  const account: EdgeAccount = useSelector(state => state.core.account)
  const clientId: string = useSelector(state => state.core.context.clientId)
  const queue: ActionQueueMap = useSelector(state => state.actionQueue.queue)

  const executionContext: ExecutionContext = useMemo(
    () => ({
      account,
      clientId
    }),
    [account, clientId]
  )

  //
  // Initialization
  //

  // @ts-expect-error
  useAsyncEffect(async () => {
    if (account?.dataStore != null) {
      const store = makeActionQueueStore(account, clientId)
      const queue = await store.getActionQueueMap()
      dispatch({
        type: 'ACTION_QUEUE/LOAD_QUEUE',
        data: queue
      })
    }
  }, [account, dispatch])

  //
  // Runtime Loop
  //

  React.useEffect(() => {
    if (queue == null) return

    const { account, clientId } = executionContext

    // Loop function
    const task = async () => {
      // Programs that require immediate attention
      const urgentProgramIds = Object.keys(queue).filter(programId => {
        const { state } = queue[programId]
        // Don't execute programs which are assigned to another client
        if (state.clientId !== clientId) return false
        // Don't execute programs which are currently executing
        if (state.executing) return false
        // Don't execute programs which have not reached their scheduled time
        if (!state.effective && state.nextExecutionTime > Date.now()) return false
        return true
      })
      // Act on urgent programs
      const promises = urgentProgramIds.map(async programId => {
        const { program, state } = queue[programId]

        // Set program state to executing
        await dispatch(
          updateActionProgramState({
            ...state,
            executing: true
          })
        )

        if (program.mockMode) {
          const { nextState } = await mockActionProgram(account, program, state)
          // Update program state
          dispatch(updateActionProgramState({ ...nextState, executing: false }))
          return
        }

        const { nextState } = await executeActionProgram(executionContext, program, state).catch((error: Error) => {
          console.warn(new Error('Action Program Exception: ' + error.message))
          console.error(error)
          return {
            nextState: {
              ...state,
              effect: { type: 'done', error },
              effective: true,
              nextExecutionTime: -1
            }
          }
        })

        // Update program state
        // @ts-expect-error
        await dispatch(updateActionProgramState({ ...nextState, executing: false }))
      })
      await Promise.all(promises)
    }

    // Setup loop
    const periodicTask = makePeriodicTask(task, EXECUTION_INTERVAL, {
      onError: error => {
        const errorMessage = String(error)
        console.warn(new Error('Unexpected error in ActionQueueService: ' + errorMessage))
        console.error(error)
      }
    })

    periodicTask.start()

    // Cleanup loop
    return () => periodicTask.stop()
  }, [dispatch, executionContext, queue])

  // Return no component/view
  return null
}
