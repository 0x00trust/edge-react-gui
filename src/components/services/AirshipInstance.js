// @flow

import { makeAirship } from '../common/Airship.js'

export const Airship = makeAirship()
