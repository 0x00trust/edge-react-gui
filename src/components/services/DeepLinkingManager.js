// @flow

import React from 'react'
import { Alert } from 'react-native'
import { Actions } from 'react-native-router-flux'
import { connect } from 'react-redux'
import { sprintf } from 'sprintf-js'
import parse from 'url-parse'

import { selectWallet } from '../../actions/WalletActions.js'
import { pluginUrlMap } from '../../constants/plugins/buySellPlugins.js'
import { PLUGIN_VIEW_DEEP, SCAN } from '../../constants/SceneKeys.js'
import s from '../../locales/strings.js'
import { type Dispatch, type State as ReduxState } from '../../types/reduxTypes.js'

type StateProps = {
  addressDeepLinkData: Object,
  deepLinkPending: boolean,
  totalWalletCount: number,
  wallets: Object
}

type DispatchProps = {
  markAddressDeepLinkDone(): void,
  selectWallet(walletId: string, currencyCode: string): void
}

type Props = StateProps & DispatchProps

class DeepLinkingManagerComponent extends React.Component<Props> {
  render () {
    return null
  }

  componentDidUpdate () {
    if (Object.keys(this.props.wallets).length > 0 && this.props.deepLinkPending) this.checkForWallet()
  }
  processPluginDeepLink = (parsedUrl: Object) => {
    if (parsedUrl.pathname.includes('simplex')) {
      const plugin = pluginUrlMap['co.edgesecure.simplex']
      Actions[PLUGIN_VIEW_DEEP]({ plugin })
      this.props.markAddressDeepLinkDone()
      return
    }
    this.props.markAddressDeepLinkDone()
  }

  checkForWallet () {
    const { addressDeepLinkData } = this.props
    const { currencyCode } = addressDeepLinkData
    // check to see what we have for a deep link.
    const parsedUrl = parse(addressDeepLinkData.uri, {}, false)
    if (parsedUrl.hostname === 'plugins') {
      this.processPluginDeepLink(parsedUrl)
      return
    }
    if (!currencyCode) {
      Actions[SCAN]()
      return
    }

    for (const wallet in this.props.wallets) {
      if (this.props.wallets[wallet].currencyCode === currencyCode) {
        this.props.selectWallet(this.props.wallets[wallet].id, currencyCode)

        Actions[SCAN]()

        return
      }
    }

    if (Object.keys(this.props.wallets).length === this.props.totalWalletCount) {
      this.props.markAddressDeepLinkDone()

      const currency = this.convertCurrencyStringFromCurrencyCode(currencyCode)
      const noWalletMessage = sprintf(s.strings.alert_deep_link_no_wallet, currency, currency)
      Alert.alert(noWalletMessage)
    }
  }

  convertCurrencyStringFromCurrencyCode (code: string) {
    switch (code) {
      case 'BTC':
        return 'Bitcoin'
      case 'BCH':
        return 'Bitcoin Cash'
      case 'ETH':
        return 'Ethereum'
      case 'LTC':
        return 'Litecoin'
      case 'DASH':
        return 'Dash'
      case 'RBTC':
        return 'RSK'
      default:
        return ''
    }
  }
}

export const DeepLinkingManager = connect(
  (state: ReduxState): StateProps => ({
    addressDeepLinkData: state.core.deepLinking.addressDeepLinkData,
    deepLinkPending: state.core.deepLinking.deepLinkPending,
    totalWalletCount: state.ui.wallets.activeWalletIds.length,
    wallets: state.ui.wallets.byId
  }),

  (dispatch: Dispatch): DispatchProps => ({
    markAddressDeepLinkDone () {
      dispatch({ type: 'ADDRESS_DEEP_LINK_COMPLETE' })
    },
    selectWallet (walletId: string, currencyCode: string) {
      dispatch(selectWallet(walletId, currencyCode))
    }
  })
)(DeepLinkingManagerComponent)
