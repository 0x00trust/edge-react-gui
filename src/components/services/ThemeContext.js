// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig.js'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
