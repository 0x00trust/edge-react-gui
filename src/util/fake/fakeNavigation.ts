import { NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener(event, callback) {
    return () => undefined
  },
  isFocused() {
    return true
  },

  navigate(name, params) {},
  push(name, params) {},
  replace(name, params) {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popTo: () => {},
  popToTop() {},

  closeDrawer() {},
  openDrawer() {},
  toggleDrawer() {},

  getState() {
    return {
      drawerOpen: false,
      index: 0,
      routes: []
    }
  }
}
