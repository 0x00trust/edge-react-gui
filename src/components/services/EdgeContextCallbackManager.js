// @flow

import type { EdgeContext } from 'edge-core-js'
import React from 'react'
import { connect } from 'react-redux'

import type { Dispatch, State } from '../../types/reduxTypes.js'
import { AlertDropdown } from '../navigation/AlertDropdown.js'
import { Airship } from './AirshipInstance.js'

type EdgeContextCallbackManagerStateProps = {
  context: EdgeContext
}

type Props = EdgeContextCallbackManagerStateProps

class EdgeContextCallbackManager extends React.Component<Props> {
  componentDidUpdate (oldProps: Props) {
    if (this.props.context && this.props.context !== oldProps.context) {
      const { context } = this.props

      context.on('error', (error: mixed) => {
        console.log(error)

        // TODO: Run the errors through our translation infrastructure:
        const message = error instanceof Error ? error.message : String(error)

        return Airship.show(bridge => <AlertDropdown bridge={bridge} message={message} warning />)
      })
    }
  }

  render () {
    return null
  }
}

const mapStateToProps = (state: State): EdgeContextCallbackManagerStateProps => ({
  context: state.core.context.context
})

const mapDispatchToProps = (dispatch: Dispatch) => ({})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EdgeContextCallbackManager)
