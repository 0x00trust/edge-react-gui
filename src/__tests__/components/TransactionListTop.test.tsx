import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

import { getTheme } from '../../components/services/ThemeContext'
import { TransactionListTopComponent } from '../../components/themed/TransactionListTop'
import { fakeNavigation } from '../../util/fake/fakeNavigation'

describe('TransactionListTop', () => {
  it('should render with loading props', () => {
    const renderer = createRenderer()

    const actual = renderer.render(
      <TransactionListTopComponent
        cryptoAmount="string"
        currencyCode="string"
        pluginId="string"
        denominationName="string"
        fiatCurrencyCode="string"
        fiatBalance="string"
        fiatSymbol="string"
        walletName="string"
        isAccountBalanceVisible
        stakingBalances={{
          FIO: {
            crypto: 'string',
            fiat: 'string'
          }
        }}
        onSelectWallet={() => undefined}
        onChangeSortingState={() => undefined}
        onSearchTransaction={() => undefined}
        toggleBalanceVisibility={() => undefined}
        walletId="string"
        isEmpty={false}
        searching={false}
        navigation={fakeNavigation}
        theme={getTheme()}
      />
    )

    expect(actual).toMatchSnapshot()
  })
})
