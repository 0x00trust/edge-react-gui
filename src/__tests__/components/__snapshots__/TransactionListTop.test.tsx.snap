// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionListTop should render 1`] = `
Array [
  <View
    style={
      Object {
        "flex": 1,
        "marginBottom": 11,
        "paddingBottom": 22,
        "paddingHorizontal": 22,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 11,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "column",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "backgroundColor": "rgba(255, 255, 255, 0)",
                "borderRadius": 11,
                "justifyContent": "center",
                "minHeight": 67,
                "paddingHorizontal": 22,
              },
              Object {
                "flexGrow": 0,
                "marginBottom": 0,
                "marginLeft": 0,
                "marginRight": 0,
                "marginTop": 0,
                "paddingLeft": 65,
                "paddingRight": 65,
                "paddingVertical": 0,
              },
            ]
          }
        >
          <View
            animatedStyle={
              Object {
                "value": Object {
                  "borderColor": "rgba(164, 199, 223, 1)",
                  "right": 11,
                },
              }
            }
            collapsable={false}
            style={
              Object {
                "borderColor": "rgba(164, 199, 223, 1)",
                "borderTopWidth": 1,
                "bottom": 0,
                "left": 22,
                "position": "absolute",
                "right": 11,
              }
            }
          />
          <View
            animatedStyle={
              Object {
                "value": Object {
                  "borderColor": "rgba(164, 199, 223, 1)",
                },
              }
            }
            collapsable={false}
            style={
              Object {
                "borderBottomLeftRadius": 11,
                "borderBottomWidth": 1,
                "borderColor": "rgba(164, 199, 223, 1)",
                "borderLeftWidth": 1,
                "borderRightWidth": 0,
                "borderTopLeftRadius": 11,
                "borderTopWidth": 1,
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "top": 0,
                "width": 22,
              }
            }
          />
          <View
            animatedStyle={
              Object {
                "value": Object {
                  "borderColor": "rgba(164, 199, 223, 1)",
                },
              }
            }
            collapsable={false}
            style={
              Object {
                "borderBottomRightRadius": 11,
                "borderBottomWidth": 1,
                "borderColor": "rgba(164, 199, 223, 1)",
                "borderLeftWidth": 0,
                "borderRightWidth": 1,
                "borderTopRightRadius": 11,
                "borderTopWidth": 1,
                "bottom": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": 22,
              }
            }
          />
          <View
            animatedStyle={
              Object {
                "value": Object {
                  "borderColor": "rgba(164, 199, 223, 1)",
                  "left": 22,
                },
              }
            }
            collapsable={false}
            style={
              Object {
                "borderColor": "rgba(164, 199, 223, 1)",
                "borderTopWidth": 1,
                "left": 22,
                "position": "absolute",
                "right": 22,
                "top": 0,
              }
            }
          />
          <View
            style={
              Array [
                Object {
                  "height": 67,
                  "justifyContent": "center",
                  "left": 0,
                  "paddingHorizontal": 22,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {
                  "paddingLeft": 65,
                  "paddingRight": 65,
                },
              ]
            }
          >
            <Text
              animatedStyle={
                Object {
                  "value": Object {
                    "color": "rgba(164, 199, 223, 1)",
                    "transform": Array [
                      Object {
                        "translateX": -0,
                      },
                      Object {
                        "translateY": -0,
                      },
                      Object {
                        "scale": 1,
                      },
                    ],
                  },
                }
              }
              collapsable={false}
              numberOfLines={1}
              onLayout={[Function]}
              style={
                Object {
                  "alignSelf": "flex-start",
                  "color": "rgba(164, 199, 223, 1)",
                  "fontFamily": "Quicksand-Regular",
                  "fontSize": 22,
                  "padding": 0,
                  "transform": Array [
                    Object {
                      "translateX": -0,
                    },
                    Object {
                      "translateY": -0,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                }
              }
            >
              Search Transactions
            </Text>
          </View>
          <Text
            animatedStyle={
              Object {
                "value": Object {
                  "opacity": 0,
                },
              }
            }
            collapsable={false}
            numberOfLines={1}
            style={
              Object {
                "bottom": -28,
                "color": "#E85466",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 17,
                "left": 28,
                "opacity": 0,
                "position": "absolute",
              }
            }
          />
          <Text
            animatedStyle={
              Object {
                "value": Object {
                  "color": "rgba(164, 199, 223, 1)",
                  "opacity": 0,
                  "transform": Array [
                    Object {
                      "translateX": 6,
                    },
                    Object {
                      "scale": 0,
                    },
                  ],
                },
              }
            }
            collapsable={false}
            numberOfLines={1}
            onLayout={[Function]}
            style={
              Object {
                "bottom": -10,
                "color": "rgba(164, 199, 223, 1)",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 17,
                "opacity": 0,
                "position": "absolute",
                "right": 28,
                "transform": Array [
                  Object {
                    "translateX": 6,
                  },
                  Object {
                    "scale": 0,
                  },
                ],
              }
            }
          >
            
          </Text>
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                Object {
                  "color": "rgba(255, 255, 255, .75)",
                  "fontSize": 22,
                  "left": 0,
                  "padding": 22,
                  "position": "absolute",
                  "top": 0,
                },
                Object {
                  "fontFamily": "anticon",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            î™°
          </Text>
          <TextInput
            autoFocus={false}
            editable={true}
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            returnKeyType="search"
            secureTextEntry={false}
            selectionColor="#FFFFFF"
            style={
              Array [
                Object {
                  "alignSelf": "stretch",
                  "color": "#FFFFFF",
                  "fontFamily": "Quicksand-Regular",
                  "fontSize": 22,
                  "padding": 0,
                },
                Object {
                  "flexGrow": 0,
                },
              ]
            }
            textAlignVertical="top"
            value=""
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "marginTop": 34,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "flexShrink": 1,
                  "marginRight": 11,
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "height": 34,
                    "marginBottom": 0,
                    "marginLeft": 0,
                    "marginRight": 0,
                    "marginTop": 0,
                    "width": 34,
                  }
                }
              >
                <RNSVGSvgView
                  align="xMidYMid"
                  bbHeight={50}
                  bbWidth={50}
                  focusable={false}
                  height={50}
                  meetOrSlice={0}
                  minX={-8}
                  minY={-8}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "borderWidth": 0,
                      },
                      Object {
                        "left": -8,
                        "position": "absolute",
                        "top": -8,
                        "transform": Array [
                          Object {
                            "rotateZ": "-90deg",
                          },
                        ],
                      },
                      Object {
                        "flex": 0,
                        "height": 50,
                        "width": 50,
                      },
                    ]
                  }
                  vbHeight={50}
                  vbWidth={50}
                  width={50}
                >
                  <RNSVGGroup
                    matrix={
                      Array [
                        1,
                        0,
                        0,
                        1,
                        0,
                        0,
                      ]
                    }
                  >
                    <RNSVGCircle
                      cx={17}
                      cy={17}
                      opacity={1}
                      propList={
                        Array [
                          "stroke",
                          "strokeWidth",
                          "strokeDasharray",
                          "strokeDashoffset",
                          "strokeLinecap",
                        ]
                      }
                      r={19}
                      stroke={4284935592}
                      strokeDasharray={
                        Array [
                          119.38052083641213,
                          119.38052083641213,
                        ]
                      }
                      strokeDashoffset={null}
                      strokeLinecap={1}
                      strokeWidth={4}
                    />
                  </RNSVGGroup>
                </RNSVGSvgView>
                <View
                  style={
                    Array [
                      Object {
                        "overflow": "hidden",
                      },
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      },
                    ]
                  }
                >
                  <FastImageView
                    resizeMode="cover"
                    source={
                      Object {
                        "uri": "https://content.edge.app/currencyIcons/bitcoin/bitcoin.png",
                      }
                    }
                    style={
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                      }
                    }
                  />
                </View>
              </View>
              <Text
                adjustsFontSizeToFit={true}
                minimumFontScale={0.65}
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "color": "#FFFFFF",
                      "fontFamily": "Quicksand-Regular",
                      "fontSize": 22,
                      "includeFontPadding": false,
                    },
                    Object {
                      "flexShrink": 1,
                      "fontSize": 28,
                      "marginLeft": 11,
                      "marginRight": 6,
                    },
                    null,
                  ]
                }
              >
                wallet name
              </Text>
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#66EDA8",
                      "fontSize": 34,
                    },
                    undefined,
                    Object {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                ïˆ”
              </Text>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignSelf": "center",
                  "opacity": 1,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#66EDA8",
                      "fontSize": 34,
                    },
                    undefined,
                    Object {
                      "fontFamily": "custom",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                î ’
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              adjustsFontSizeToFit={true}
              minimumFontScale={0.3}
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "Quicksand-Regular",
                    "fontSize": 22,
                    "includeFontPadding": false,
                  },
                  Object {
                    "fontFamily": "Quicksand-Medium",
                    "fontSize": 45,
                  },
                  null,
                ]
              }
            >
              0.00123123 BTC
            </Text>
            <Text
              adjustsFontSizeToFit={true}
              minimumFontScale={0.65}
              numberOfLines={1}
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "Quicksand-Regular",
                    "fontSize": 22,
                    "includeFontPadding": false,
                  },
                  Object {
                    "fontSize": 28,
                  },
                  null,
                ]
              }
            >
              $0.00 USD
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 22,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "flexShrink": 1,
            "justifyContent": "flex-start",
            "opacity": 1,
            "paddingRight": 34,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#66EDA8",
                "fontSize": 22,
              },
              undefined,
              Object {
                "fontFamily": "anticon",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          î˜™
        </Text>
        <Text
          adjustsFontSizeToFit={true}
          minimumFontScale={0.65}
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "includeFontPadding": false,
              },
              Object {
                "color": "#66EDA8",
                "fontFamily": "Quicksand-Medium",
                "fontSize": 22,
                "marginLeft": 6,
              },
              null,
            ]
          }
        >
          Request
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "flexShrink": 1,
            "justifyContent": "flex-start",
            "opacity": 1,
            "paddingRight": 34,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#66EDA8",
                "fontSize": 22,
              },
              undefined,
              Object {
                "fontFamily": "anticon",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          î˜š
        </Text>
        <Text
          adjustsFontSizeToFit={true}
          minimumFontScale={0.65}
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "includeFontPadding": false,
              },
              Object {
                "color": "#66EDA8",
                "fontFamily": "Quicksand-Medium",
                "fontSize": 22,
                "marginLeft": 6,
              },
              null,
            ]
          }
        >
          Send
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginBottom": -22,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 22,
          "opacity": 1,
          "paddingBottom": 11,
          "paddingLeft": 11,
          "paddingRight": 11,
          "paddingTop": 11,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(255, 255, 255, 0)",
            "borderColor": "rgba(255, 255, 255, .1)",
            "borderWidth": 1,
            "flexDirection": "row",
            "padding": 11,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "height": 45,
                "margin": 11,
                "width": 45,
              },
            ]
          }
        >
          <FastImageView
            resizeMode="contain"
            source={
              Object {
                "uri": "https://content.edge.app/currencyIcons/bitcoin/bitcoin.png",
              }
            }
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
        </View>
        <Text
          adjustsFontSizeToFit={true}
          minimumFontScale={0.65}
          numberOfLines={0}
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "includeFontPadding": false,
              },
              Object {
                "flex": 1,
                "margin": 11,
              },
              null,
            ]
          }
        >
          Spend USD -&gt; Earn BTC
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginLeft": 22,
          "paddingBottom": 22,
        },
        null,
        undefined,
      ]
    }
  >
    <Text
      adjustsFontSizeToFit={true}
      minimumFontScale={0.65}
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Quicksand-Regular",
            "fontSize": 22,
            "includeFontPadding": false,
          },
          Object {
            "fontFamily": "Quicksand-Medium",
          },
          null,
        ]
      }
    >
      Transactions
    </Text>
  </View>,
  <View
    style={
      Object {
        "marginBottom": 11,
        "marginLeft": 22,
      }
    }
  >
    <BVLinearGradient
      colors={
        Array [
          452984831,
          452984831,
        ]
      }
      endPoint={
        Object {
          "x": 1,
          "y": 0.5,
        }
      }
      locations={null}
      startPoint={
        Object {
          "x": 0,
          "y": 0.5,
        }
      }
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "height": 1,
          },
          Object {
            "marginBottom": 0,
            "marginLeft": 0,
            "marginRight": -22,
            "marginTop": 0,
          },
        ]
      }
    />
  </View>,
]
`;
